### PR TITLE
Improve TypeScript definition

### PR DESCRIPTION
**Pull request by marcogrcr**
WHAT?

1. Fixed bugs in definitions:
   - Renamed `AwsSigner` to `AwsV4Signer`.
   - Removed `retries` and `initRetryMs` fields from: `AwsSignerInit`.
   - Removed `body`, `headers`, `method`, and `url` fields from:
     `AwsRequestInit["aws"]`.
   - Made all fields optional in `AwsRequestInit["aws"]`.
   - Removed invalid types (`Blob`, `FormData`, `URLSearchParams`,
     `ReadableStream`) from `AwsSignerInit["body"]`.
   - Removed optional suffix (`?`) from required fields.
   - Removed redundant `| null` type in optional fields.
2. Made all `interface` fields `readonly`.
3. Added `authHeader()` and `signature()` to `AwsV4Signer`.
4. Added `jsdoc` to the TypeScript definitions based on the
   `README.md` documentation.
5. Added `dtslint`-based tests.

WHY?

1. Allow the library to be used correctly with TypeScript.
2. Make the interfaces immutable.
3. Allow invocation of missing public methods.
4. Make the library easier to use.
5. Ensures types are correct and changes are backwards compatible.

### CODE REVIEW

File: types/test.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
+import { AwsClient, AwsRequestInit, AwsV4Signer } from "aws4fetch";
```
Comment (marcogrcr) [inline types/test.ts:1]: Previous types don't allow `AwsV4Signer` to be imported.

File: types/test.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+function AwsClientTest() {
+  function constructor() {
+    // $ExpectError
+    new AwsClient({});
```
Comment (marcogrcr) [inline types/test.ts:6]: Previous types fail this test.

File: types/test.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+function AwsV4SignerTest() {
+  function constructor() {
+    // $ExpectError
+    new AwsV4Signer();
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      retries: 1, // $ExpectError
```
Comment (marcogrcr) [inline types/test.ts:98]: Previous types fail this test.

File: types/test.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      initRetryMs: 1, // $ExpectError
```
Comment (marcogrcr) [inline types/test.ts:105]: Previous types fail this test.

File: types/test.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new Blob(), // $ExpectError
```
Comment (marcogrcr) [inline types/test.ts:112]: Previous types fail this test.

File: types/test.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new FormData(), // $ExpectError
```
Comment (marcogrcr) [inline types/test.ts:119]: Previous types fail this test.

File: types/test.ts [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new URLSearchParams(), // $ExpectError
```
Comment (marcogrcr) [inline types/test.ts:126]: Previous types fail this test.

File: types/test.ts [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new ReadableStream(), // $ExpectError
```
Comment (marcogrcr) [inline types/test.ts:133]: Previous types fail this test.

File: types/test.ts [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+      datetime: "value",
+      appendSessionToken: true,
+      allHeaders: true,
+      singleEncode: true,
+    });
+  }
+
+  function authHeader(signer: AwsV4Signer) {
+    // $ExpectType Promise<string>
+    signer.authHeader();
```
Comment (marcogrcr) [inline types/test.ts:163]: Previous types don't allow this method to be called.

File: types/test.ts [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      body: new ArrayBuffer(1),
+    }).sign();
+
+    // $ExpectType Promise<AwsRequestInfo<Uint8Array>>
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new Uint8Array(),
+    }).sign();
```
Comment (marcogrcr) [inline types/test.ts:196]: Previous types don't provide generics support for getting a properly types `body`.

File: types/test.ts [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+  }
+
+  function signature(signer: AwsV4Signer) {
+    // $ExpectType Promise<string>
+    signer.signature();
```
Comment (marcogrcr) [inline types/test.ts:201]: Previous types don't allow this method to be called.

**Comment by mhart** [comment]
Hi @marcogrcr – thanks for the PR!

I went with the solution I said I was exploring here: https://github.com/mhart/aws4fetch/pull/7#issuecomment-631709172

I think this should cover all the issues you had. I'll close this, but happy to reopen if you think there's something missing

**Comment by marcogrcr** [comment]
I'm happy to adjust my pull request to include the missing:

- Declaration testing with dtslint for ensuring compatibility and non-breaking changes.
- `jsdoc`.

Let me know if you see value in that.

**Comment by mhart** [comment]
I don't think I'll add dslint until they remove tslint and move to eslint

But happy to look at any jsdoc changes you'd like to make 👍

**Pull request closed without merging**